Border agent

LeZhang2016 opened the pull request:
- Achieved the function of the Border Agent
- Fixed some issues

xiaom-GitHub commented on src/core/coap/secure_coap_server.cpp:
```diff
@@ -117,7 +117,8 @@ void SecureServer::Receive(Message &aMessage, const Ip6::MessageInfo &aMessageIn
         mPeerAddress.SetPeerAddr(aMessageInfo.GetPeerAddr());
         mPeerAddress.SetPeerPort(aMessageInfo.GetPeerPort());
-
+        mPeerAddress.SetSockAddr(aMessageInfo.GetSockAddr());
```
Do we also need to enable macro "OPENTHREAD_ENABLE_BORDER_AGENT" here? If we do not enable border agent, maybe we do not need to set source address and port.

LeZhang2016 replied:
In Thread Network, both the Commissioner and Border Agent need to start secure CoAP server, so the Commissioner may also face the same problem, not just the Border Agent.

xiaom-GitHub commented on src/core/thread/thread_netif.hpp:
```diff
@@ -299,6 +303,12 @@ class ThreadNetif: public Ip6::Netif
 
+#if OPENTHREAD_ENABLE_BORDER_AGENT
+    MeshCoP::BorderAgent &GetBorderAgent(void) { return mBorderAgent; }
+
```
extra line?

LeZhang2016 replied:
I will remove this line, Thanks!

LeZhang2016 replied:
I have checked the code again, and i followed the code style of the  OPENTHREAD_ENABLE_COMMISSIONER and OPENTHREAD_ENABLE_JOINER

#if OPENTHREAD_ENABLE_COMMISSIONER
    MeshCoP::Commissioner &GetCommissioner(void) { return mCommissioner; }

    Coap::SecureServer &GetSecureCoapServer(void) { return mSecureCoapServer; }
#endif  // OPENTHREAD_ENABLE_COMMISSIONER

#if OPENTHREAD_ENABLE_BORDER_AGENT
    MeshCoP::BorderAgent &GetBorderAgent(void) { return mBorderAgent; }

    Coap::SecureServer &GetBASecureCoapServer(void) { return mBASecureCoapServer; }
#endif  // OPENTHREAD_ENABLE_BORDER_AGENT

#if OPENTHREAD_ENABLE_DTLS
    MeshCoP::Dtls &GetDtls(void) { return mDtls; }
#endif  // OPENTHREAD_ENABLE_DTLS

#if OPENTHREAD_ENABLE_JOINER
    MeshCoP::Joiner &GetJoiner(void) { return mJoiner; }

    Coap::SecureClient &GetSecureCoapClient(void) { return mSecureCoapClient; }
#endif  // OPENTHREAD_ENABLE_JOINER

xiaom-GitHub commented on src/core/thread/thread_uris.hpp:
```diff
@@ -40,224 +40,240 @@ namespace Thread {
  */
-#define OPENTHREAD_URI_DIAGNOSTIC_RESET       "d/dr"
+#define OPENTHREAD_URI_DIAGNOSTIC_RESET         "d/dr"
 
```
move below URIs macro here?

xiaom-GitHub commented on src/core/thread/thread_uris.hpp:
```diff
@@ -40,224 +40,240 @@ namespace Thread {
  */
-#define OPENTHREAD_URI_COMMISSIONER_SET       "c/cs"
+#define OPENTHREAD_URI_COMMISSIONER_SET         "c/cs"
 
```
move below URIs macro here? in alphabetical order.

LeZhang2016 replied:
I will adjust the position of the URI according to the alphabetical order. Thanks.

bukepo commented on src/core/meshcop/border_agent.cpp:
```diff
@@ -0,0 +1,600 @@
+{
+    ThreadError error = kThreadError_Error;
+    PSKcTlv *pskc;
+    char aThreadPSKc[] = 
```
I think aThreadPSKc is not needed, since `SetPsk()` has prototype `ThreadError Dtls::SetPsk(const uint8_t *aPsk, uint8_t aPskLength)`.  Just give it `pskc->GetPSKc()`.

LeZhang2016 replied:
What you've said is reasonable here but, i want a variable to record the pskc default value, if the border agent cannot obtain the pskc from the Leader (data response tlv), it can use the default value to start DTLS server.

bukepo replied:
Is it reasonable to start border agent if PSKc is not available?

LeZhang2016 replied:
PSKc is used to authenticate the External Commissioner or Native Commissioner. According to the spec, the border agent should obtain the available PSKc when it join into the Thread Network. But in OT, i think we should configure the PSKc after we start the border agent.

bukepo commented on src/core/meshcop/border_agent.cpp:
```diff
@@ -0,0 +1,600 @@
+    }
+
+    SuccessOrExit(error = mNetif.GetSecureCoapServer().SetPsk(reinterpret_cast<const uint8_t *>(aThreadPSKc),
+                                                              static_cast<uint8_t>(sizeof(aThreadPSKc))));
```
`sizeof(aThreadPSKc)` may be changed to a meaningful macro or const int value for it's fixed for Thread.

bukepo commented on src/core/meshcop/border_agent.hpp:
```diff
@@ -0,0 +1,152 @@
+    Timer mTimer;
+    ThreadNetif &mNetif;
+    Ip6::Address mCommissionerAddr;
+    uint8_t mCommissionerUdpPort;
```
This should be uint16_t?

LeZhang2016 replied:
I will change it. Thanks

ZhangLe2016 committed:
Achieve the functions of Border Agent

- Establish the commissioning session with External Commissioner via DTLS
- Handle Commissioner Petitioning (COMM_PET, LEAD_PET) and keep the commissioning session open (COMM_KA, LEAD_PET)
- Handle Commissioner Manangement (MGMT_SET)
- Relay Joiner authentication messages (RLY_RX, RLY_TX) between the External Commissioner and Joiner Router

ZhangLe2016 committed:
Enable the MAC layer security to support the message forward on Border Agent

ZhangLe2016 committed:
Solve issue: The default size of application data is short for Border Agent relaying "RLY_RX" message

Border Agent relays the RLY_RX to External Commissioner over Commissioning Session. The RLY_RX message is packaged in DTLS application data. If the RLY_RX wraps the "DTLS-ClientHello", the size of whole message is more than 500 bytes.`

jwhui committed:
Remove private `code_utils.hpp` include from examples. (#1534)

jwhui committed:
Update CONTRIBUTING.md (#1546)

jwhui committed:
Use fast poll after sending an MLE Data Request. (#1551)

abtink committed:
`Ncp`: Simplify handling of NCP instance object from callbacks and C APIs (#1552)

This commit makes the following changes:

- It adds a static class member variable `sNcpInstance` and a static
  getter method `GetNcpInstance()` to `NcpBase` class.

- It updates `NcpBase` and its sub-classes `NcpUart` and `NcpSpi` to
  use the new method from the callbacks and C APIs (combining/replacing
  3 separate static variables tracking the NCP instance).

jwhui committed:
Make VerifyOrExit() action argument optional. (#1554)

bukepo committed:
choose ephemeral port at udp layer (#1556)

aeliot committed:
Parent last rssi (#1558)

* Added thread api call to get the last rssi for the parent

wbober committed:
Always build platform utils library. (#1557)

This commit changes build behaviour, so that platform-utils library
is always built. This allows for using opentread in environments which
keep platform files out of openthread tree.

bukepo commented on src/core/meshcop/border_agent.cpp:
```diff
@@ -0,0 +1,596 @@
+void BorderAgent::HandleLeaderMgmtCommissionerSetResponse(void *aContext, otCoapHeader *aHeader, otMessage *aMessage,
+                                                          const otMessageInfo *aMessageInfo, ThreadError aResult)
+{
+    static_cast<BorderAgent *>(aContext)->HandleLeaderPetitionResponse(
```
typo here. should be
```
    static_cast<BorderAgent *>(aContext)->HandleLeaderMgmtCommissionerSetResponse(
        *static_cast<Coap::Header *>(aHeader), *static_cast<Message *>(aMessage),
        *static_cast<const Ip6::MessageInfo *>(aMessageInfo), aResult);
```

LeZhang2016 replied:
@bukepo, 
Thanks a lot for your support!!

bukepo reviewed (commented):
found a typo

bukepo commented on src/core/meshcop/border_agent.cpp:
```diff
@@ -0,0 +1,596 @@
+
+    otLogFuncEntry();
+    otLogCritMeshCoP(GetInstance(), "received petition from commissioner");
+
```
mInstance.GetSecureCoapServer().SendEmptyAck(aHeader, aMessageInfo);

bukepo commented on src/core/meshcop/border_agent.cpp:
```diff
@@ -0,0 +1,596 @@
+
+    otLogFuncEntry();
+    otLogCritMeshCoP(GetInstance(), "received keep alive from commissioner");
+
```
mInstance.GetSecureCoapServer().SendEmptyAck(aHeader, aMessageInfo);

bukepo commented on src/core/meshcop/border_agent.cpp:
```diff
@@ -0,0 +1,596 @@
+    otLogFuncEntry();
+    otLogCritMeshCoP(GetInstance(), "received MGMT_COMMISSIONER_SET request from Commissioner");
+    otLogCritMeshCoP(GetInstance(), "send MGMT_COMMISSIONER_SET request to Leader");
+
```
mInstance.GetSecureCoapServer().SendEmptyAck(aHeader, aMessageInfo);

LeZhang2016 replied:
@bukepo, I will add them according to your comments. Thanks.

bukepo reviewed (changes requested):
Please add these lines to send empty ack to client. This stops client retries.

ZhangLe2016 committed:
To avoid the CoAPS request retransmission from the Extermal Commissioner, Border Agent replys the empty ACK after accepting the confirmable CoAPS request(COMM-PET, COMM-KA, MGMT-SET req) from EC.

jwhui committed:
Remove Active and Pending Datasets from unsolicited Data Response. (#1237)

bukepo committed:
pass to host without hop limit decrease (#1562)

librasungirl committed:
not configure discovery scan duration (#1502)

aeliot committed:
Moved the cca error counter so that it will count even when using radio csma rather then OT csma (#1566)

librasungirl committed:
CC2650: enable bootloader backdoor (#1555)

abtink committed:
Skip performing "make pretty" on copied `strlcpy.c` and `strlcat.c` (#1568)

jwhui committed:
Remove use of zero-length arrays. (#1570)

abtink committed:
Adding `SourceMatchController` class. (#1559)

This commit adds a new class `SourceMatchController` which
contains all (existing) logic to maintain the source match table
and control when/how to enable or disable the source matching
feature.

It also updates `Child` class by changing the access level of
member variables related to the source matching feature (e.g.,
number of indirect queued messages) to private to ensure that
they are only modified by the new `SourceMatchController` class
while providing simple getter methods for use by other classes.

The changes in this commit also address two corner-case issues
related to source address matching feature:

- First issue where we could possibly enable the source address
  matching after successfully adding a short address while still
  having remaining pending entries.

- Second issue where an entry would not be added properly from a
  `SetSrcMatchAsShort()` call for a child, if there were more
  than one message queued for the child.

abtink committed:
`DataPollManager`: Update some of the methods/variables names. (#1561)

bukepo committed:
Separate response support (#1571)

* Move empty message to coap_base.

Both client and server needs this feature to handle separate
response.

* remove unnecessary check

* make CoapBase::SendEmptyMessage() private.

* add comment to indicate intentional fall through

rlubos committed:
Assert otPlatUartSend errors in ncp_uart (#1563)

bukepo committed:
disable commissioner related CoAP resource before accepted (#1573)

* disable commissioner related coap before accepted

LuDuda committed:
Remove usage of CoAP kAckTimeout in Link Raw API. (#1580)

Gr3yh0und committed:
[CoAP] Adding CLI commands (#1539)

* [CoAP] Adding CLI commands

Adds CLI commands for a simple CoAP client and server with one resource

bukepo committed:
fix a wrong usage of spinel_datatype_unpack (#1575)

jwhui committed:
Add accessor methods to neighbor/child/router objects. (#1577)

chshu committed:
Use a new specialized ChildInfo structure for children information persistent storage (#1572)

Create a new specialized ChildInfo structure, which will be only used for children information storage

Also update otPlatSettingsGet() to be consistent with its documentation:
* return the actual length of the setting, but not the actual length it read to aValue

abtink committed:
`ip6`: Ensure not to pass a message from local host back to it. (#1583)

The commit makes the following changes:

- We ensure not to pass up a message we received from local host
  back to the host again (this avoids message loops due to incorrect
  route setup on the host).

- We add logs to indicate when the process of passing up a message to
  host fails (e.g., running out of message buffers and thus failing
  to clone the message).

chshu committed:
Refine parent information storage (#1578)

librasungirl committed:
1) Make the rebooted child able to sync the link with the parent which has also experienced reset (#1579)

2) Network Data TLV is not always included

jwhui committed:
Update CC2538 flashing instructions. (#1582)

chshu committed:
Fix incorrect logic in IncrementMessageCount() method (#1589)

librasungirl committed:
Parent keeps restoring children until success or timeout (#1581)

bukepo committed:
Refine API for creating CoAP message (#1584)

* merge two CoAP api for new message

* decouple CoAP and MeshCoP

* create a wrapper for creating meshcop message

* create meshcop.hpp

jwhui committed:
Do not revert to non-volatile settings when changed by application. (#1586)

- Invalidate Active and Pending Datasets when configs are changed via APIs.
- Do not allow client to change network configurations while Thread is enabled.

abtink committed:
`spinel.h`: Adjust spacing and alignments (again...) (#1587)

This commit fixes the spacing and alignment in `spinel.h` header
file. This was done earlier in commit 227d2a091b but was overwritten
(back to the older version) in commit 2fa7a13c1b (which intended to add
two new spinel properties related to commissioning in `spinel.h`).

Vyrastas committed:
Revised README / Landing page (#1585)

* Updating README draft

* Complete rewrite
* Split logos into individual images
* Added new images for codelab and ncp

ZhangLe2016 committed:
Sync up with Master branch

ZhangLe2016 committed:
Change the logic "wait for the ChildID response to get PSKc"

if the PSK is invalid, the CLI outputs failed infomation.

ZhangLe2016 committed:
change the code style by "make pretty" operation
